pkg/graveler/graveler.go

```diff
@@ -1070,29 +1074,65 @@ func (g *KVGraveler) UpdateBranch(ctx context.Context, repositoryID RepositoryID
 	if reference.ResolvedBranchModifier == ResolvedBranchModifierStaging {
 		return nil, fmt.Errorf("reference '%s': %w", ref, ErrDereferenceCommitWithStaging)
 	}
-	newBranch := Branch{}
 
-	err = g.RefManager.BranchUpdate(ctx, repositoryID, branchID, func(curBranch *Branch) (*Branch, error) {
+	if err := g.prepareForCommitIDUpdate(ctx, repositoryID, branchID, repo); err != nil {
+		return nil, err
+	}
+
+	var tokensToDrop []StagingToken
+	var newBranch *Branch
+	err = g.RefManager.BranchUpdate(ctx, repositoryID, branchID, func(currBranch *Branch) (*Branch, error) {
 		// TODO(Guys) return error only on conflicts, currently returns error for any changes on staging
-		empty, err := g.stagingEmpty(ctx, repositoryID, repo, curBranch)
+		empty, err := g.sealedEmpty(ctx, repositoryID, repo, currBranch)
 		if err != nil {
 			return nil, err
 		}
 		if !empty {
 			return nil, ErrConflictFound
 		}
 
-		newBranch = *curBranch
-		newBranch.CommitID = reference.CommitID
-		return &newBranch, nil
+		tokensToDrop = currBranch.SealedTokens
+		currBranch.SealedTokens = []StagingToken{}
+		currBranch.CommitID = reference.CommitID
+		newBranch = currBranch
+		return currBranch, nil
 	})
 	if err != nil {
 		if errors.Is(err, kv.ErrPredicateFailed) {
 			err = ErrConflictFound
 		}
 		return nil, err
 	}
-	return &newBranch, nil
+	g.dropTokens(ctx, tokensToDrop...)
+
+	return newBranch, nil
+}
+
+// prepareForCommitIDUpdate will check that the staging area is empty,
+// before adding a new staging token. It is best to use it before changing
+// the branch HEAD as a preparation for deleting the staging area.
+// see issue #3771 for more information on the algorithm used.
+func (g *KVGraveler) prepareForCommitIDUpdate(ctx context.Context, repositoryID RepositoryID, branchID BranchID, repo *Repository) error {
```

Comment:
It's currently not needed as it will always be followed up by another `BranchUpdate`, so using the returned branch from here could result with errors.